chainstate/src/detail/mod.rs

```diff
@@ -271,52 +274,179 @@ impl<S: BlockchainStorage, V: TransactionVerificationStrategy> Chainstate<S, V>
+
+        let reorg_occurred = chainstate_ref.activate_best_chain(block_index, &best_block_id)?;
+        block_status.advance_validation_stage_to(BlockValidationStage::FullyChecked);
+
+        Ok(reorg_occurred)
+    }
+
+    // Attempt to process the block. On success, return Some(block_index_of_the_passed_block)
+    // if a reorg has occurred and the passed block is now the best block, otherwise return None.
+    fn attempt_to_process_block(
+        &mut self,
+        block: WithId<Block>,
+        block_source: BlockSource,
+    ) -> Result<Option<BlockIndex>, BlockError> {
+        let block = self.check_legitimate_orphan(block_source, block).log_err()?;
+        let block_id = block.get_id();
+
+        // Ensure that the block being submitted is new to us. If not, bail out immediately,
+        // otherwise create a new block index and continue.
+        let block_index = {
+            let chainstate_ref = self.make_db_tx_ro().map_err(BlockError::from).log_err()?;
+            let existing_block_index = chainstate_ref
+                .get_block_index(&block_id)
+                .map_err(BlockError::BlockLoadError)
+                .log_err()?;
+
+            if let Some(block_index) = existing_block_index {
+                return if block_index.status().is_valid() {
+                    Err(BlockError::BlockAlreadyProcessed(block_id))
+                } else {
+                    Err(BlockError::InvalidBlockAlreadyProcessed(block_id))
+                };
+            }
+
+            chainstate_ref.new_block_index(&block, BlockStatus::new()).log_err()?
+        };
+
+        // Perform block checks; `check_result` is `Result<bool>`, where the bool indicates
+        // whether a reorg has occurred.
+        let (block_status, check_result) = self.with_rw_tx_and_state_for_block_id(
+            |chainstate_ref, block_status| {
+                Self::perform_block_checks(chainstate_ref, &block, &block_index, block_status)
+            },
+            &BlockStatus::new(),
+            &block_id,
+            "block validation",
+        );
+
+        if let Err(err @ BlockError::DatabaseCommitError(_, _, _, _)) = check_result {
+            // If we got here, then the block checks have succeeded, but the DB has failed.
+            // Attempts to save the new status in BlockIndex in this situation will
+            // probably fail too. Moreover, even if we succeed, we'll get a strange situation
+            // where there is a BlockIndex in the DB with a "fully checked" status, but the
+            // block itself is missing. So we bail out in this case.
+            return Err(err);
         }
+
+        let block_index = {
+            let mut block_index = block_index;
+            block_index.set_status(block_status);
+            block_index
+        };
+
+        // Update block index status.
+        let status_update_result = self.with_rw_tx_for_block_id(
```

Comment:
wouldn't it be simpler to set block status inside `perform_block_checks` to make it one transaction?

Reply:
aah right, missed that